src/server/editorServices.ts

```diff
@@ -463,7 +448,45 @@ namespace ts.server {
                     this.typingsCache.deleteTypingsForProject(response.projectName);
                     break;
             }
-            project.updateGraph();
+            project.markAsDirty();
+            this.delayUpdateProjectGraphAndInferredProjectsRefresh(project);
+        }
+
+        private delayInferredProjectsRefresh() {
+            this.pendingInferredProjectUpdate = true;
+            this.throttledOperations.schedule("*refreshInferredProjects*", /*delay*/ 250, () => {
+                if (this.pendingProjectUpdates.size !== 0) {
+                    this.delayInferredProjectsRefresh();
+                }
+                else if (this.pendingInferredProjectUpdate) {
+                    this.pendingInferredProjectUpdate = false;
+                    this.refreshInferredProjects();
+                }
+            });
+        }
+
+        private delayUpdateProjectGraph(project: Project) {
+            const projectName = project.getProjectName();
+            this.pendingProjectUpdates.set(projectName, project);
+            this.throttledOperations.schedule(projectName, /*delay*/ 250, () => {
+                const project = this.pendingProjectUpdates.get(projectName);
+                if (project) {
+                    this.pendingProjectUpdates.delete(projectName);
+                    project.updateGraph();
+                }
+            });
+        }
+
+        delayUpdateProjectGraphAndInferredProjectsRefresh(project: Project) {
+            this.delayUpdateProjectGraph(project);
+            this.delayInferredProjectsRefresh();
+        }
+
+        private delayUpdateProjectGraphs(projects: Project[]) {

```

Comment:
When this is called with `projects = this.inferredProjects` (at the end of the method immediately below this), doesn't `delayInferredProjectsRefresh` update all the inferred projects anyway? Wouldn't that update twice?